### PR TITLE
python310Packages.numba: Backport support for numpy 1.24.x

### PR DESCRIPTION
**Pull request by mweinelt**
This is a last ditch effort to keep numba working.

The patch is 13 kB large, but this is the price we have to pay.

This  is numba/numba/pull/8691 sans documentation and ci changes.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by FRidh** [comment]
If it works and upstream does not report any major issues, then it is OK for me.

**Comment by mweinelt** [comment]
Has been reviewed and they are waiting on some package availability, not sure.

https://github.com/numba/numba/pull/8691#pullrequestreview-1299282164

**Comment by fabaff** [comment]
Result of `nixpkgs-review pr 218929` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages marked as broken and skipped:</summary>
  <ul>
    <li>intensity-normalization</li>
    <li>python310Packages.bambi</li>
    <li>python310Packages.caffe</li>
    <li>python310Packages.caffeWithCuda</li>
    <li>python310Packages.graspologic</li>
    <li>python310Packages.imgaug</li>
    <li>python310Packages.intensity-normalization</li>
    <li>python310Packages.mask-rcnn</li>
    <li>python310Packages.pygbm</li>
    <li>python310Packages.pywick</li>
    <li>python310Packages.qiskit</li>
    <li>python310Packages.qiskit-machine-learning</li>
    <li>python310Packages.rl-coach</li>
    <li>python310Packages.rlax</li>
    <li>python310Packages.shap</li>
  </ul>
</details>
<details>
  <summary>86 packages failed to build:</summary>
  <ul>
    <li>easyocr (python310Packages.easyocr)</li>
    <li>home-assistant-component-tests.environment_canada</li>
    <li>larynx-train (python310Packages.larynx-train)</li>
    <li>napari (python310Packages.napari)</li>
    <li>ome_zarr (python310Packages.ome-zarr)</li>
    <li>pianotrans</li>
    <li>pitivi</li>
    <li>python310Packages.aeppl</li>
    <li>python310Packages.aesara</li>
    <li>python310Packages.aplpy</li>
    <li>python310Packages.arviz</li>
    <li>python310Packages.awkward</li>
    <li>python310Packages.batchgenerators</li>
    <li>python310Packages.boxx</li>
    <li>python310Packages.bpycv</li>
    <li>python310Packages.bsuite</li>
    <li>python310Packages.casa-formats-io</li>
    <li>python310Packages.clifford</li>
    <li>python310Packages.cppe</li>
    <li>python310Packages.dask</li>
    <li>python310Packages.dask-gateway</li>
    <li>python310Packages.dask-glm</li>
    <li>python310Packages.dask-image</li>
    <li>python310Packages.dask-jobqueue</li>
    <li>python310Packages.dask-ml</li>
    <li>python310Packages.dask-mpi</li>
    <li>python310Packages.dask-yarn</li>
    <li>python310Packages.datashader</li>
    <li>python310Packages.devito</li>
    <li>python310Packages.distributed</li>
    <li>python310Packages.dm-sonnet</li>
    <li>python310Packages.dremel3dpy</li>
    <li>python310Packages.env-canada</li>
    <li>python310Packages.fastparquet</li>
    <li>python310Packages.ffcv</li>
    <li>python310Packages.galois</li>
    <li>python310Packages.glymur</li>
    <li>python310Packages.gumath</li>
    <li>python310Packages.handout</li>
    <li>python310Packages.hyppo</li>
    <li>python310Packages.image-match</li>
    <li>python310Packages.imagecorruptions</li>
    <li>python310Packages.imageio</li>
    <li>python310Packages.intake</li>
    <li>python310Packages.intake-parquet</li>
    <li>python310Packages.librosa</li>
    <li>python310Packages.moviepy</li>
    <li>python310Packages.napari-console</li>
    <li>python310Packages.napari-svg</li>
    <li>python310Packages.numbaWithCuda</li>
    <li>python310Packages.phik</li>
    <li>python310Packages.piano-transcription-inference</li>
    <li>python310Packages.pims</li>
    <li>python310Packages.psd-tools</li>
    <li>python310Packages.pyfftw</li>
    <li>python310Packages.pygmo</li>
    <li>python310Packages.pymc</li>
    <li>python310Packages.pymoo</li>
    <li>python310Packages.pynndescent</li>
    <li>python310Packages.pyscf</li>
    <li>python310Packages.pytensor</li>
    <li>python310Packages.pyvista</li>
    <li>python310Packages.runway-python</li>
    <li>python310Packages.scikit-tda</li>
    <li>python310Packages.scikitimage</li>
    <li>python310Packages.sfepy</li>
    <li>python310Packages.slicedimage</li>
    <li>python310Packages.sparse</li>
    <li>python310Packages.spectral-cube</li>
    <li>python310Packages.streamz</li>
    <li>python310Packages.stumpy</li>
    <li>python310Packages.stytra</li>
    <li>python310Packages.sunpy</li>
    <li>python310Packages.tensorflow-datasets</li>
    <li>python310Packages.test-tube</li>
    <li>python310Packages.tifffile</li>
    <li>python310Packages.torchlibrosa</li>
    <li>python310Packages.trackpy</li>
    <li>python310Packages.umap-learn</li>
    <li>python310Packages.uproot</li>
    <li>python310Packages.vector</li>
    <li>python310Packages.xarray-einstats</li>
    <li>python310Packages.zcs</li>
    <li>tartube</li>
    <li>tartube-yt-dlp</li>
    <li>tts</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.numba</li>
    <li>python310Packages.numba-scipy</li>
    <li>python310Packages.resampy</li>
    <li>python310Packages.timezonefinder</li>
  </ul>
</details>

**Comment by mweinelt** [comment]
Let's roll with it and see if we can make things work.
